File: benchmarks/cut.pl
#!/usr/bin/env perl

# This is a very simplistic version of 'cut'. It was written for performance comparisons,
# it is not intended for real work.

use 5.012;
use feature 'unicode_strings';
use open IO => ':encoding(utf8)';
use open ':std';

use Getopt::Long;
use strict;

&main();

sub help {
    my ($ostream) = @_;
    if (!defined($ostream)) { $ostream = *STDOUT; }
    print $ostream <<EOF;
Synopis: $0 [options] -- [file ...]

Options:
   --h|help                Print this help.
   --f|fields N [N ...]    Fields to cut. Specified as 1-upped indices.

EOF
}

sub main {
    my $b_help = 0;
    my @field_indices = ();

    my $r = GetOptions
      (
       'h|help!'     => \$b_help,
       'f|fields=i{,}'  => \@field_indices,
      );

    if (!$r) {
        &help(*STDERR);
        exit(1);
    } elsif ($b_help) {
        &help();
        return;
    }

    @field_indices = map { $_ - 1 } @field_indices;

    while (<>) {
        chomp();
        say join("\t", (split("\t", $_))[@field_indices]);
    }
}

File: benchmark_code/cut.py
#!/usr/bin/env python

# This is a very simplistic version of 'cut'. It was written for performance comparisons,
# it is not intended for real work.

import argparse
import fileinput

def main():
    parser = argparse.ArgumentParser(description='Simple cut utility in python.')
    parser.add_argument('-f', '--fields', nargs='+', type=int, required=True)
    parser.add_argument('files', nargs='*')

    args = parser.parse_args();
    field_indicies = [x - 1 for x in args.fields]

    for line in fileinput.input(args.files):
        fields = line.rstrip('\n').split('\t')
        print '\t'.join([fields[x] for x in field_indicies])
    
if __name__ == '__main__':
    main()
    
